tools/souper-interpret.cpp

```diff
@@ -33,6 +34,35 @@ static cl::opt<unsigned> DebugLevel("souper-debug-level", cl::init(1),
      "The larger the number is, the more fine-grained debug "
      "information will be printed."));
 
+static cl::list<std::string>
+AbstractInterpreters("abstract-interps", cl::desc("list of abstract interpreters to run"),
+		     cl::CommaSeparated);
+
+namespace {
+  enum class CompareDataflowResult {
+    SAME,
+    LESS,
+    GREATER,
+    INCOMPARABLE
+  };
+
+  CompareDataflowResult compareKnownBits(KnownBits &a, KnownBits &b) {
+    assert(!a.hasConflict() && !b.hasConflict() && "Can't compare conflicting KnownBits operands!");
+
+    APInt unknownA = ~(a.One ^ a.Zero);
+    APInt unknownB = ~(b.One ^ b.Zero);
+    if ((unknownA | unknownB).countPopulation() >
+	std::max(unknownA.countPopulation(), unknownB.countPopulation()))
+      return CompareDataflowResult::INCOMPARABLE;
```

Comment:
I don't think this is right: two values conflict when one of them thinks a bit is 1 and the other thinks its 0

I would encourage you to implement "compare" for known bits and then just use the same code to compare known bits values as you use for constantranges

Reply:
yeah, you are right. i missed a case here.

Reply:
pushed a modified version. Now, I am checking for the conflicts first in the known bits and after we are clear that there is no conflict in known bits, we are moving forward with the code I had earlier.